### DISCUSSION TITLE
RFC: Dependencies

### DISCUSSION THREAD

**Discussion by hseg** [papis/papis#779] [General]
I'm annotating the `papis-git` [AUR package](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=papis-git) with the dependencies for the tests (and which test needs what). I hope I got it right, would appreciate a review.
In particular, am unsure whether I got the `flake8` deps right.

Also, notice the extended discussion on why `jinja`, `markdownify`, and `whoosh` are optional for the `pytest` tests, as explained in https://github.com/papis/papis/issues/776#issuecomment-1991874151.
Given this optionality, and the fact that jinja and whoosh are supposed to be plugins, perhaps it might be possible to decouple them further, as is done for `papis-libgen` and `papis-scihub`? Not sure if it's too big an ask, though

**Comment by alexfikl** [comment]
It looks good to me! Few extras:
* `python-sphinx-click`: should only be needed when building docs?
* `pylint`: this isn't really used in our CI or anything at the moment. We wanted to integrate it at some point, but haven't gotten around to finishing that up (it throws a lot of errors :(
* `python-pytest-mypy`: this shouldn't be needed if mypy is ran directly.

Then for the dependencies
* `jinja2` is pretty nice, since the default formatter is very limited. I'm more inclined to make it a full dependency.
* `whoosh` has a good chance of being phased out, since it doesn't seem very maintained and the performance isn't great. Maybe we can replace it with an `sqlite` backend? :grin:
* `markdownify` is very new, so not sure. We might want something to clean various pieces of text that we scrape off the web and this seems pretty nice (it's also just one small file on top of `bs4`).

I don't have a very strong opinion about the optdeps, so maybe someone else has some other thoughts.

**Comment by alexfikl** [comment]
Also, since we were talking about it in #607, maybe it would be cool to include the man pages? Mostly to see if they can be packaged easily. Something like this should work
```
  # build man pages
  python -m build --wheel --no-isolation .
  python -m installer --destdir=tmp_man_install dist/*.whl

  local site_packages=$(python -c 'import site; print(site.getsitepackages()[0])')
  export PYTHONPATH="$(pwd)/tmp_man_install/${site_packages}:${PYTHONPATH}"
  make -C doc man

  # rebuild wheel to contain man pages
  python -m build --wheel --no-isolation .
```

**Comment by hseg** [comment]
OK, so I tested these build instructions, and they indeed work, see the latest
version of the PKGBUILD I just uploaded.

As for python-sphinx-click -- I understood it was only necessary to build the
docs, that's why I tagged it as a makedepends (necessary only at build time) and
not depends (necessary at run time), checkdepends (necessary at test time) or
optdepends (optional at run time).

**Comment by hseg** [comment]
Oh, and killed the pylint and python-pytest-mypy deps as remarked.

As for jinja2, whoosh, and markdownify -- the annoyance is less that they _are_
optional dependencies, and more that they are semi-mandatory since they're
needed to run the tests.

What I had in mind was more that the coupling for them be loose enough that they
could be shipped as independent plugins, and to move the integration tests to
those plugins. Of course, besides the added complexity of needing to pluginify
these interfaces, this would also incur codebase fragmentation, so I'm not sure
to what extent it's worth it.

The alternative of baking these deps in and have the coupling only be loose
enough that the _project_ can switch libraries easily, but it's not expected
that users will, is also a valid way of addressing this concern, and is IIUC
your intended approach?

**Reply by alexfikl** [reply]
`jinja2` and `whoosh` can definitely be shipped as external plugins, since both the formatter and the database backends already have a plugin architecture. `markdownify` is a bit less flexible, but the full Zenodo downloader can itself be shipped as an external plugin.

Is this really a problem though? You can just add them to `checkdepends`, no?
Also, the `pytest` tests should run fine without them, so it's just the `mypy` tests that would fail, right?

**Comment by hseg** [comment]
P.S. As you'll see in the PKGBUILD, the project builds fine with
`python -m build` as well. I'm not sure what value-add hatchling provides for
you guys, but given that I'm justifying running mypy and flake8 tests on the
grounds that the semantics of a `-git` package is the bleeding-edge development
package, I'll move it over if you find it necessary.

**Reply by alexfikl** [reply]
`hatching` is the build backend and `python -m build` is the build frontend, so I would expect them to just work together. You will have to add `python-hatchling` to the `makedepends` of the PKGBUILD though (it shouldn't build without it in a clean chroot now).

There's a lot of discussion in #607 about `hatchling`. We mostly went with it over staying with `setuptools` because it seemed to have less bugs around editable installs and reproducible packages.

**Comment by hseg** [comment]
Yeah, it's probably more trouble than it's worth. I've indeed added them to checkdepends, and have instructed users to edit the PKGBUILD if they don't need the plugin.
It's the reverse, though - mypy tests have all the typing deps, pytest requires these plugin dependencies to be there for the integration tests.

**Comment by hseg** [comment]
Ah, oops. I haven't been very good about building in clean chroots, will fix it tomorrow.
